### PR TITLE
Add EMR Serverless Operators and Hooks

### LINKED ISSUE #20215: EMR serverless, new operator
### Description

A new EMR serverless has been announced and it is already available, see:

- https://aws.amazon.com/blogs/big-data/announcing-amazon-emr-serverless-preview-run-big-data-applications-without-managing-servers/
- https://aws.amazon.com/emr/serverless/

Having an operator for creating applications and submitting jobs to EMR serverless would be awesome.

### Use case/motivation

New operator for working with EMR serverless.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by syedahsn**
Closes: #20215

Define new operators, as well as necessary hooks and sensors, to run EMR Serverless jobs. System test for EMR Serverless will be coming in a separate PR.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
@ferruzzi @o-nikolas @vincbeck -> WDYT?

**Comment by vincbeck** [comment]
Syed is a new member of our team, we already reviewed his work :) Thanks for reaching out though

**Comment by potiuk** [comment]
> Syed is a new member of our team, we already reviewed his work :) Thanks for reaching out though

aha :) Good to know.

File: airflow/providers/amazon/aws/operators/emr.py
Comment (potiuk) [inline airflow/providers/amazon/aws/operators/emr.py]: I tihnk we are already fine with our util class (as in other providers) since we moved to 2.2+

**Comment by potiuk** [comment]
Some errors to fix :)

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/hooks/emr.py]:
```suggestion
        :param failure_states: A set of states which indicate failure and should throw an
            exception if any are reached before the desired_state
        :param object_type: Used for the reporting string. What are you waiting for? (application, job, etc)
        :param action: Used for the reporting string. What action are you waiting for? (created, deleted, etc)
        :param countdown: Total amount of time the waiter should wait for the desired state
            before timing out (in seconds). Defaults to 25 * 60 seconds.
        :param check_interval_seconds: Number of seconds waiter should wait before attempting
            to retry get_state_callable. Defaults to 60 seconds.
```
Would you mind add some indentation? The docstrings typically have a hanging indent in parameter descriptions on lines beyond the first for readability.

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/hooks/emr.py]: Might be better to use `logging` instead of `print`? Again, just another typical practice.

File: airflow/providers/amazon/aws/operators/emr.py
Comment (josh-fell) [reply airflow/providers/amazon/aws/operators/emr.py]: +1 You can import `cached_property` from [`airflow.compat.functools`](https://github.com/apache/airflow/blob/main/airflow/compat/functools.py).

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]:
```suggestion
from airflow.compat.functools import cached_property
```
Same idea here as above. You can use the native compat util for `cached_property`

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]:
```suggestion
    :param target_states: a set of states to wait for, defaults to 'SUCCESS'
```
Maybe it makes sense to use the actual value being checked rather than the variable in the docstring. I was thinking users reading the Python API docs won't have to look at code or other places in the docs to figure out what `SUCCESS_STATES` is. WDYT?

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]: @vincbeck Is this something you can answer?

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]:
```suggestion
        return EmrServerlessHook(aws_conn_id=self.aws_conn_id)
```
So users can control the connection ID used to build the hook object. Unless this should be something like  use `aws_conn_id` if provided or `emr_conn_id` otherwise?

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]:
```suggestion
        return EmrServerlessHook(aws_conn_id=self.aws_conn_id)
```

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]: I could easily be missing it, but where is `emr_conn_id` being used in this sensor and other sensors in this file?

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/sensors/emr.py]:
```suggestion
    :param target_states: a set of states to wait for, defaults to {'STARTED', 'CREATED'}
```
Same idea about having the values in the docstring. Up to you.

File: airflow/providers/amazon/aws/operators/emr.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/emr.py]: Sounds good, I'll make the change

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/sensors/emr.py:213]: This was a question posed to one of the internal reviewers. This should be removed.

Comment (syedahsn) [reply airflow/providers/amazon/aws/sensors/emr.py]: This was a question posed to one of the internal reviewers. I'll remove it from here.

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/emr.py]: Same here I guess :)

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/sensors/emr.py]: Agree

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/sensors/emr.py]: I agree. It seems `emr_conn_id` is not used. You can just remove it

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/sensors/emr.py]: I dont think this line is needed

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/sensors/emr.py]: Overall question. Do we want or is there any value converting the `Exception` to `AirflowException`? I can see there is a lot of operators which do not do that. If there is no value, I would rather not doing the conversion just for the sake of having a code cleaner

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/sensors/emr.py]: Agree, we should pass `aws_conn_id`

File: docs/apache-airflow-providers-amazon/operators/emr_serverless.rst
Comment (vincbeck) [inline docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:33]:
```suggestion
Operators
---------
```

File: docs/apache-airflow-providers-amazon/operators/emr_serverless.rst
Comment (vincbeck) [inline docs/apache-airflow-providers-amazon/operators/emr_serverless.rst]:
```suggestion
==========================
```

**Review by vincbeck — CHANGES_REQUESTED** [review]

**Comment by dacort** [comment]
Closes https://github.com/apache/airflow/issues/20215

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/sensors/emr.py]: I went with `Exception` just as a catchall for any type of exception that might be thrown by the boto3 API. I'm not sure if there's an advantage to throwing an `AirflowException` instead of a generic `Exception`. Only thing I can think of is that it would make it clear that the exception was coming from Airflow, rather than any underlying API calls (i.e. boto3) that might have been made. But that can also be understood from the message associated with the `Exception`.

**Comment by potiuk** [comment]
some errors need fixing.

File: docs/apache-airflow-providers-amazon/operators/emr_serverless.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-amazon/operators/emr_serverless.rst]: This underline needs to be at least the length of the text it's underlining. I imagine this might be a Build Docs failure for you.

File: airflow/providers/amazon/aws/operators/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/operators/emr.py]:
```suggestion
        if app_state not in EmrServerlessApplicationSensor.SUCCESS_STATES:
```
Small suggestion to use the attrs where you can like you have in other places. Up to you though (assuming it's functionally correct as well).

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (josh-fell) [reply airflow/providers/amazon/aws/sensors/emr.py]: FWIW I think `AirflowException` is used too broadly in the code base. IMO `AirflowException`, and its subclasses, are really related to Airflow orchestration activity rather than activity done by what Airflow is _orchestrating_. It might make more sense to drop the `try/except` and bubble up whatever API call errors that might be thrown directly. Up to you though.

**Review by josh-fell — APPROVED** [review]
Some minor suggestions but overall LGTM (after fixing the tests of course).

File: airflow/providers/amazon/aws/sensors/emr.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/sensors/emr.py]: certainly makes to code cleaner.  :+1:

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (josh-fell) [inline airflow/providers/amazon/aws/hooks/emr.py]: FYI - A couple of `print()` statements here.

File: docs/apache-airflow-providers-amazon/operators/emr_serverless.rst
Comment (vincbeck) [inline docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:35]:
```suggestion

Operators
---------

```

**Review by vincbeck — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/example_dags/example_emr_serverless.py
Comment (eladkal) [inline airflow/providers/amazon/aws/example_dags/example_emr_serverless.py:1]: new example dags should be according to AIP-47
Example: https://github.com/apache/airflow/pull/25205

**Review by eladkal — DISMISSED** [review]

File: airflow/providers/amazon/aws/example_dags/example_emr_serverless.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/example_dags/example_emr_serverless.py:1]: That's a good point but the plan was to do it in two iterations since this PR is already big enough. Once this PR is merged, @syedahsn will work on converting this sample dag to system test. If you strongly disagree, I guess the two can be done in this PR but it will just make it bigger and harder to review

**Pull request merged**
